docs/writing/posts/retrieval-is-more-than-pure-chance.md

```diff
@@ -0,0 +1,675 @@
+
+```
+>>> np.linspace(1,0,4)
+array([1.        , 0.66666667, 0.33333333, 0.        ])
+```
+
+Since we only have a single relevant chunk, only our known chunk will have a label of `1` while all of the others will have a value of 0. 
+
+#### MRR
+
+MRR refers to Mean Reciprocal Rank is much simpler to implement and understand. It's important to note here that the MRR is only useful in calculating the retrieval quality with respect to the **first relevant item in the returned list**. If we have multiple relevant items, then NDCG is a much better alternative.
+
+If we have 4 items `a,b,c,d` where only `a` is the relevant item, then our Reciprocal Rank(RR) is simply 1. If `c` is the relevant item,then our RR is going to be `1/3`. But, if all of the items are irrelevant, then our RR is going to be 0
+
+ We can express this in the form of a function as 
+
+```python
+def calculate_mrr(chunk_id:str,predictions:List[str]):
+    return 0 if chunk_id not in predictions else 1 / (predictions.index(chunk_id) + 1)
+```
+
+Once we've performed this calculation for a few different retrievals, then the MRR is simply the mean of the individual RRs for each list of retrieved chuns.
+
+### Calculation
+
+Now, let's bring it all together in a single call 
+
+```python
+results = [["a", "b", "c"], ["b", "a", "c"], ["b", "c", "a"]]
+
+evals = {
+    "MRR":calculate_mrr,
+    "NDCG":calculate_ndcg
+}
+evaluations = [
+    {metric: fn("a", result) for metric, fn in evals.items()} for result in results
+]
+
+df = pd.DataFrame(evaluations)
+console = Console()
+console.print(df)
+```
+
+This gives us the values as seen below
+
+
+|MRR |     NDCG |
+| --- | --- |
+| 1.00 |  1.00 |
+| 0.50 |  0.63 |
+| 0.33 |  0.50 |
+
+
+As we can see, as the relevant item ( in this example above 'a' ) gets shifted further, our MRR and NDCG decrease slowly. But, MRR decreases significantly faster and more aggressively than that of NDCG.
+
+### Varying K
+
+Now, we also want to make sure we can calculate these metrics for different values of `k` - to do so, we can use a simple decorator which passes a slice of the list of elements into the metric calculation. 
+
+```python
+def slice_predictions_decorator(num_elements: int):
+    def decorator(func):
+        def wrapper(chunk_id, predictions):
+            sliced_predictions = predictions[:num_elements]
+            return func(chunk_id, sliced_predictions)
+
+        return wrapper
+
+    return decorator
+```
+
+Let's scale our original example just now to 5 values where `a` gets shifted back progressively by one position on each iteration.
+
+
+```python
+evals = {}
+SIZES = [3, 5]
+for size in SIZES:
+    evals[f"MRR@{size}"] = slice_predictions_decorator(size)(calculate_mrr)
```

Comment:
why is this needed to be a decorator 

slice_predictions_decorator(size)(calculate_mrr) is only ever used twice 

just do 

def calculate_mrr(data, k)
  return score(data[:k]